### ISSUE TITLE
Poor Performance on Multi-Round RefCOCO with Unchanging Training Loss

### ISSUE THREAD

**Issue by lmsdss** [berkeley-hipie/segllm#9]
Hi,
Thank your for this interesting work!

I'm encountering issues with training performance. Despite using multiple GPUs (4) and a batch size of 2 per device, the results on the Multi-Round RefCOCO dataset are not satisfactory. The evaluation metrics are showing poor performance (P@0.5, mIoU, oIoU) for each round, and the training loss seems to stabilize without much improvement over time.

Evaluation Results:
Here are the evaluation results for each round:

Round 0: P@0.5 = 0.404, mIoU = 0.357, oIoU = 0.253
Round 1: P@0.5 = 0.391, mIoU = 0.348, oIoU = 0.240
Round 2: P@0.5 = 0.385, mIoU = 0.347, oIoU = 0.248
Round 3: P@0.5 = 0.368, mIoU = 0.332, oIoU = 0.239
Round 4: P@0.5 = 0.346, mIoU = 0.304, oIoU = 0.217
Round 5: P@0.5 = 0.326, mIoU = 0.290, oIoU = 0.206
Round 6: P@0.5 = 0.229, mIoU = 0.209, oIoU = 0.138
Round 7: P@0.5 = 0.254, mIoU = 0.228, oIoU = 0.142
Round 8: P@0.5 = 0.287, mIoU = 0.262, oIoU = 0.178
Round 9: P@0.5 = 0.160, mIoU = 0.143, oIoU = 0.089
Round 10: P@0.5 = 0.236, mIoU = 0.216, oIoU = 0.197
Round 11: P@0.5 = 0.105, mIoU = 0.102, oIoU = 0.098
Round 12: P@0.5 = 0.308, mIoU = 0.256, oIoU = 0.137
Round 13: P@0.5 = 0.300, mIoU = 0.283, oIoU = 0.247
Round 14: P@0.5 = 0.091, mIoU = 0.074, oIoU = 0.024
Round 15: P@0.5 = 0.333, mIoU = 0.269, oIoU = 0.123
Round 16: P@0.5 = 0.0, mIoU = 0.0, oIoU = 0.0

Training Loss:
Here are the training loss details at different steps:

Step 12000:
{'loss': 37.5, 'grad_norm': 374.51971435546875, 'learning_rate': 1.3683285563217418e-06, 'epoch': 0.8}
{'loss': 32.875, 'grad_norm': 176.00950622558594, 'learning_rate': 1.3682285567540795e-06, 'epoch': 0.8}
{'loss': 42.6875, 'grad_norm': 255.4354248046875, 'learning_rate': 1.3681285529263784e-06, 'epoch': 0.8}
{'loss': 36.8125, 'grad_norm': 182.24073791503906, 'learning_rate': 1.3680285448397953e-06, 'epoch': 0.8}
{'loss': 36.875, 'grad_norm': 164.53233337402344, 'learning_rate': 1.3679285324954874e-06, 'epoch': 0.8}

Step 30000:
{'loss': 33.2188, 'grad_norm': 216.4274139404297, 'learning_rate': 7.12700245917297e-11, 'epoch': 1.99}
{'loss': 44.5625, 'grad_norm': 482.6876525878906, 'learning_rate': 6.999167935217708e-11, 'epoch': 1.99}
{'loss': 47.4375, 'grad_norm': 287.866943359375, 'learning_rate': 6.872490231057959e-11, 'epoch': 1.99}
{'loss': 38.875, 'grad_norm': 249.56979370117188, 'learning_rate': 6.74696934813701e-11, 'epoch': 1.99}
{'loss': 35.5312, 'grad_norm': 323.8143310546875, 'learning_rate': 6.622605287909255e-11, 'epoch': 1.99}

The loss remains almost unchanged from step 12000 to step 30000, even though the learning rate has dropped to extremely low values (~10^-11).

Parameter Changes:

--eval_steps: I believe is not affecting training accuracy directly, but it does consume a lot of time. The performance might not improve significantly during the evaluation steps, so I was wondering if this could be contributing to the overall issue.

Other hyperparameters have remained unchanged.

Could anyone provide insight into why the loss is not improving and the evaluation metrics are poor?

Thanks!

**Comment by lmsdss** [comment]
Even with 8 GPUs and a batch size of 2 per device, the results on the Multi-Round RefCOCO dataset are still not good.

**GATHERED** Eval Result for Round 0: (Step 0): {'global_step': 0, 'eval/P@0.5_round0': np.float64(0.39733333333330684), 'eval/mIoU_round0': np.float64(0.35276131184893483), 'eval/oIoU_round0': np.float64(0.24695866587331894)}
**GATHERED** Eval Result for Round 1: (Step 0): {'global_step': 0, 'eval/P@0.5_round1': np.float64(0.3713333333333086), 'eval/mIoU_round1': np.float64(0.33096722412107166), 'eval/oIoU_round1': np.float64(0.22571631262529301)}
**GATHERED** Eval Result for Round 2: (Step 0): {'global_step': 0, 'eval/P@0.5_round2': np.float64(0.3788395904436537), 'eval/mIoU_round2': np.float64(0.34201549751356936), 'eval/oIoU_round2': np.float64(0.24147988719396676)}
**GATHERED** Eval Result for Round 3: (Step 0): {'global_step': 0, 'eval/P@0.5_round3': np.float64(0.3826185101579703), 'eval/mIoU_round3': np.float64(0.3468249732161361), 'eval/oIoU_round3': np.float64(0.2499719010096613)}
**GATHERED** Eval Result for Round 4: (Step 0): {'global_step': 0, 'eval/P@0.5_round4': np.float64(0.3066037735848574), 'eval/mIoU_round4': np.float64(0.2682169308452064), 'eval/oIoU_round4': np.float64(0.1851461092502358)}
**GATHERED** Eval Result for Round 5: (Step 0): {'global_step': 0, 'eval/P@0.5_round5': np.float64(0.27980535279798546), 'eval/mIoU_round5': np.float64(0.25771842269705564), 'eval/oIoU_round5': np.float64(0.19268864882925946)}
**GATHERED** Eval Result for Round 6: (Step 0): {'global_step': 0, 'eval/P@0.5_round6': np.float64(0.21495327102793696), 'eval/mIoU_round6': np.float64(0.1959416220120379), 'eval/oIoU_round6': np.float64(0.12919239852978756)}
**GATHERED** Eval Result for Round 7: (Step 0): {'global_step': 0, 'eval/P@0.5_round7': np.float64(0.272189349112265), 'eval/mIoU_round7': np.float64(0.24680364484603093), 'eval/oIoU_round7': np.float64(0.18528462820182479)}
**GATHERED** Eval Result for Round 8: (Step 0): {'global_step': 0, 'eval/P@0.5_round8': np.float64(0.252173913043259), 'eval/mIoU_round8': np.float64(0.2306875643520837), 'eval/oIoU_round8': np.float64(0.154441593624133)}
**GATHERED** Eval Result for Round 9: (Step 0): {'global_step': 0, 'eval/P@0.5_round9': np.float64(0.17333333333310222), 'eval/mIoU_round9': np.float64(0.1640954589841562), 'eval/oIoU_round9': np.float64(0.10052675741079459)}
**GATHERED** Eval Result for Round 10: (Step 0): {'global_step': 0, 'eval/P@0.5_round10': np.float64(0.3090909090903471), 'eval/mIoU_round10': np.float64(0.2979021245777609), 'eval/oIoU_round10': np.float64(0.30473471620857573)}
**GATHERED** Eval Result for Round 11: (Step 0): {'global_step': 0, 'eval/P@0.5_round11': np.float64(0.1315789473680748), 'eval/mIoU_round11': np.float64(0.12220174387849471), 'eval/oIoU_round11': np.float64(0.06744480954119232)}
**GATHERED** Eval Result for Round 12: (Step 0): {'global_step': 0, 'eval/P@0.5_round12': np.float64(0.30769230769112427), 'eval/mIoU_round12': np.float64(0.22060165038390778), 'eval/oIoU_round12': np.float64(0.1046371771323022)}
**GATHERED** Eval Result for Round 13: (Step 0): {'global_step': 0, 'eval/P@0.5_round13': np.float64(0.34999999999825004), 'eval/mIoU_round13': np.float64(0.3309380769713068), 'eval/oIoU_round13': np.float64(0.29326150081788066)}
**GATHERED** Eval Result for Round 14: (Step 0): {'global_step': 0, 'eval/P@0.5_round14': np.float64(0.09090909090826446), 'eval/mIoU_round14': np.float64(0.0850923603223521), 'eval/oIoU_round14': np.float64(0.11333505324762959)}
**GATHERED** Eval Result for Round 15: (Step 0): {'global_step': 0, 'eval/P@0.5_round15': np.float64(0.0), 'eval/mIoU_round15': np.float64(0.0), 'eval/oIoU_round15': np.float64(0.0)}
**GATHERED** Eval Result for Round 16: (Step 0): {'global_step': 0, 'eval/P@0.5_round16': np.float64(0.0), 'eval/mIoU_round16': np.float64(0.0), 'eval/oIoU_round16': np.float64(0.0)}

**Comment by marlo-z** [comment]
Hi, can you try again now? Please pull updated code.
We have updated the config file. It appears the old training config has an unexpected "test3" flag.
This flag will train a 2 layer model with 128 hidden size. It is used for single GPU debuging/testing of the training code on a very small model.
https://github.com/berkeley-hipie/segllm/blob/e17dffa36b86ef2c01faf2b72264b2c39f8f5596/llava/train/train.py#L1565-L1569

**Comment by lmsdss** [comment]
Thanks for the update!
 Is --eval_steps 5 mandatory? It seems time-consuming. Can I increase it to save time?

**Comment by marlo-z** [comment]
Of course! You can freely set the --eval_steps parameter. It is just for monitoring the validation performance during training. 
Thank you for your interested in our work!
